### DISCUSSION TITLE
Using async/await in Emscripten libraries

### DISCUSSION THREAD

**Discussion by iliakan** [emscripten-core/emscripten#18730] [Ideas]
I consider improving things in Emscripten JS.

But I see promises everywhere.

Nowadays, async/await are supported everywhere, can I use them instead of `Promise`? Please.

I have a solid argument: browsers/platforms that don't support async/await also don't have WASM.
So there won't be compatibility problems.

**Comment by sbc100** [comment]
Sadly we don't get even allow full ES6 constructors in JS libraries, at least no in general.     IIRC we do still have some folks who target very old embedded browsers and they still want to be able to target, for example, ES5!    I believe such folks use `-sWASM=0` to convert the wasm to ES5.

For more info on the transition to ES5 to ES6 see https://github.com/emscripten-core/emscripten/issues/11984

**Reply by kripken** [reply]
Can't closure lower ES6 constructors to ES5? Or perhaps Babel can? (If so, perhaps one of them also lowers async-await?)

**Reply by iliakan** [reply]
Most people use Babel for that.

But `async/await` becomes really ugly when rewritten by Babel with helper functions using a finite-state machine-like approach. 

So the downside is: if people will have to debug generated and lowered JS in the environment without async/await support, they'll have to deal with quite ugly mess instead of "async/await".

The upside is: all the others enjoy async/await and modern JS.
